Network scoped alias support

mavenugo opened the pull request:
Part #2 of the proposal : #18699 which brings in the `network scoped alias` support.

This PR + the local alias support (https://github.com/docker/docker/pull/19229) fixes #18699

mavenugo committed:
Network scoped alias support

Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo commented:
@cpuguy83 @vdemeester @calavera @icecrime  I have temporarily included the engine-api vendoring commit in this PR to help CI succeed. I will rebase this PR once we have an officially vendored in the engine-api dependency.

We didn't cover the UX portion during the maintainers meeting. Hoping to get some consensus in this PR.

tiborvass commented on runconfig/opts/parse.go:
```diff
@@ -100,6 +101,7 @@ func Parse(cmd *flag.FlagSet, args []string) (*container.Config, *container.Host
 	cmd.Var(&flVolumes, []string{"v", "-volume"}, "Bind mount a volume")
 	cmd.Var(&flTmpfs, []string{"-tmpfs"}, "Mount a tmpfs directory")
 	cmd.Var(&flLinks, []string{"-link"}, "Add link to another container")
+	cmd.Var(&flAliases, []string{"-network-alias"}, "Add network scoped alias for the container")
```
I suggest renaming this to `--net-alias` to make it more convenient and consistent with `--net`

tiborvass replied:
also the variable name should be `flNetAliases`

tiborvass commented on integration-cli/docker_cli_network_unix_test.go:
```diff
@@ -1008,3 +1008,41 @@ func verifyIPAddresses(c *check.C, cName, nwname, ipv4, ipv6 string) {
+
+func (s *DockerSuite) TestUserDefinedNetworkConnectDisconnectAlias(c *check.C) {
+	testRequires(c, DaemonIsLinux, NotUserNamespace)
+	dockerCmd(c, "network", "create", "-d", "bridge", "aliasnet1")
```
i find the name `aliasnet1` to be confusing, just call it `net1` unless I'm missing something. Same for aliasnet2

tiborvass commented:
UX wise:
I like `docker network connect ... --alias myalias`
and like I said, I would prefer changing `docker run --network-alias foo ...` to `docker run --net-alias foo ...`.

Also, we should make it very clear the relationship between `--net-alias` and `--name`.

Other than that, you have my +1.

cpuguy83 commented:
What happens if you are connected to a network already and you want to use an additional alias?

mavenugo commented:
@cpuguy83 we dont have the support to dynamically update the network scoped alias. the only way is to disconnect and reconnect with the list of new aliases.
(also, one can use --link as a way to workaround such cases where it can be handled as a consumer side alias).

cpuguy83 commented:
Also seems possible for multiple containers to specify the same alias on a network.
What's the expected behavior here, and is it documented?

mavenugo commented:
@cpuguy83 yes. it is expected (and is a requirement for compose). The expected behavior in 1.10 is to provide high availability between the containers sharing the alias in that network. When one container goes down another one will start serving the request. We are not providing built-in load-balancing yet.
It is documented and I will push the docs once I get to move past the engine-api vendoring rebase.

cpuguy83 commented:
design +1 from me as well with the changes mentioned by @tiborvass 
Once you push docs and flag changes will move to code-review.

mavenugo commented:
@cpuguy83 @tiborvass rebased on top of engine-api v0.2.1 and addressed your comments.

The docs has some dependencies on  https://github.com/docker/docker/pull/19229. Once it is merged, I will rebase this PR again to finish up the docs. PTAL.

tiborvass commented on runconfig/opts/parse.go:
```diff
@@ -100,6 +101,7 @@ func Parse(cmd *flag.FlagSet, args []string) (*container.Config, *container.Host
 	cmd.Var(&flVolumes, []string{"v", "-volume"}, "Bind mount a volume")
 	cmd.Var(&flTmpfs, []string{"-tmpfs"}, "Mount a tmpfs directory")
 	cmd.Var(&flLinks, []string{"-link"}, "Add link to another container")
+	cmd.Var(&flAliases, []string{"-net-alias"}, "Add network scoped alias for the container")
```
@mavenugo `network-scoped`

tiborvass commented on runconfig/errors.go:
```diff
@@ -33,4 +33,6 @@ var (
 	// ErrUnsupportedNetworkNoSubnetAndIP conflict between network with no configured subnet and preferred ip address
 	ErrUnsupportedNetworkNoSubnetAndIP = fmt.Errorf("User specified IP address is supported only when connecting to networks with user configured subnets")
+	// ErrUnsupportedNetworkAndAlias conflict between network mode and alias
+	ErrUnsupportedNetworkAndAlias = fmt.Errorf("Network scoped alias is supported only for containers in user defined networks")
```
`Network-scoped`

jfrazelle commented:
LGTM

mavenugo commented:
@tiborvass @jfrazelle @cpuguy83 rebased and addressed comments. PTAL.

tiborvass commented on api/client/network.go:
```diff
@@ -115,6 +115,8 @@ func (cli *DockerCli) CmdNetworkConnect(args ...string) error {
 	flLinks := opts.NewListOpts(runconfigopts.ValidateLink)
 	cmd.Var(&flLinks, []string{"-link"}, "Add link to another container")
+	flAliases := opts.NewListOpts(nil)
+	cmd.Var(&flAliases, []string{"-alias"}, "Add network scoped alias for the container")
```
`network-scoped`

tiborvass commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,128 @@ environment variables into a running container without significant effort and he
 it is not compatible with `docker network` which provides a dynamic way to connect/
 disconnect containers to/from a network.
 
+### Network scoped alias
```
`Network-scoped`

tiborvass commented on runconfig/errors.go:
```diff
@@ -33,4 +33,6 @@ var (
 	// ErrUnsupportedNetworkNoSubnetAndIP conflict between network with no configured subnet and preferred ip address
 	ErrUnsupportedNetworkNoSubnetAndIP = fmt.Errorf("User specified IP address is supported only when connecting to networks with user configured subnets")
+	// ErrUnsupportedNetworkAndAlias conflict between network mode and alias
+	ErrUnsupportedNetworkAndAlias = fmt.Errorf("Network-scoped alias is supported only for containers in user defined networks")
```
`user-defined`

tiborvass replied:
nvm, the other errors also have the spelling mistakes

mavenugo commented:
@tiborvass 

> i thought you'd grep it ;)

this time I grep'ed and caught 2 more in the comments :)... fixed now.

tiborvass commented:
code lgtm, but haven't tested yet

mavenugo commented:
ping @cpuguy83  @thaJeztah

cpuguy83 commented:
One thing that's a little weird, a container that is not explicitly named can not be resolved by an alias.

mavenugo commented:
@cpuguy83 that's correct. Its because we don't consider a container for name resolution if the name is not specified. I have to think through the cases if it make sense to support --net-alias for anonymous containers. TBH I cant think of anything obvious. Shall we treat it as a bug-fix in libnetwork and get it resolved ?

thaJeztah commented:
Hm good one, @cpuguy83. @mavenugo yes, I think aliases for anonymous containers should be supported.

mavenugo commented:
@thaJeztah @cpuguy83 i agree as well : https://github.com/docker/libnetwork/issues/871
Since the fix is entirely in libnetwork and has no change on this PR, I think we can move ahead with this PR. the next vendor in from libnetwork will solve it.

cpuguy83 commented:
@mavenugo I think treating it as a bug is ok, same with `--link` functionality also.

cpuguy83 commented:
LGTM moving to docs-review

tiborvass commented:
Ping @thaJeztah @vdemeester @jamtur01 @MHBauer

jamtur01 commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,128 @@ environment variables into a running container without significant effort and he
+### Network-scoped alias
+
+While `links` provide private name resolution that is localized within a container,
+`network alias` provides a way for a container to be discovered by an alternate name
```
the \`network alias\` provides... ?

jamtur01 commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,128 @@ environment variables into a running container without significant effort and he
+While `links` provide private name resolution that is localized within a container,
+`network alias` provides a way for a container to be discovered by an alternate name
+by any other container within the scope of a particular network.
+network alias is defined by the container that is offering the service to the network
```
The \`network alias\` is defined....

jamtur01 commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,128 @@ environment variables into a running container without significant effort and he
+(while `link` alias is defined by the consumer of a service)
+
+Continuing with the above example, create another container in `isolated_nw` with a
+network alias
```
Add .

jamtur01 commented on docs/userguide/networking/work-with-networks.md:
````diff
@@ -499,6 +499,128 @@ environment variables into a running container without significant effort and he
+round-trip min/avg/max = 0.070/0.081/0.097 ms
+```
+
+let us connect container6 to the `local_alias` network with a different `network alias`.
````
Now let us connect \`container6\` to the ...

jamtur01 commented on docs/userguide/networking/work-with-networks.md:
````diff
@@ -499,6 +499,128 @@ environment variables into a running container without significant effort and he
+$ docker network connect --alias scoped-app local_alias container6
+```
+
+The container6 in this example now is aliased as `app` in network `isolated_nw` and
````
`container6`

jamtur01 commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,128 @@ environment variables into a running container without significant effort and he
+The container6 in this example now is aliased as `app` in network `isolated_nw` and
+as `scoped-app` in network `local_alias`.
+
+Lets try to reach these aliases from container4 (which is connected to both these networks)
```
Let's

jamtur01 replied:
`container4`

jamtur01 commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,128 @@ environment variables into a running container without significant effort and he
+as `scoped-app` in network `local_alias`.
+
+Lets try to reach these aliases from container4 (which is connected to both these networks)
+and container5 (which is connected only to isolated_nw)
```
`container5`
    `isolated_nw`

jamtur01 commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,128 @@ environment variables into a running container without significant effort and he
+those containers that are connected to that network can access the alias.
+
+In addition to the above features, multiple containers can share the same Network-scoped
+alias within the same network. For example, lets launch container7 in `isolated_nw` with
```
`container7`

jamtur01 commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,128 @@ environment variables into a running container without significant effort and he
+
+In addition to the above features, multiple containers can share the same Network-scoped
+alias within the same network. For example, lets launch container7 in `isolated_nw` with
+the same alias as container6
```
`container6`

jamtur01 commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,128 @@ environment variables into a running container without significant effort and he
+
+When multiple containers share the same alias, name resolution to that alias will happen
+to one of the containers (typically the first container that is aliased). When the container
+that backs the alias goes down or disconnected from the network, next container that
```
the next container

jamtur01 commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,128 @@ environment variables into a running container without significant effort and he
+backs the alias will be resolved.
+
+Let us ping the alias `app` from container4 and bring down container6 to verify that
+container7 is resolved for the `app` alias.
```
`container7` is resolving the `app` alias.

mavenugo commented:
thanks @jamtur01 . addressed the comments.

tiborvass commented:
@jamtur01 do you approve? :)

thaJeztah commented on docs/userguide/networking/work-with-networks.md:
````diff
@@ -499,6 +499,129 @@ environment variables into a running container without significant effort and he
+round-trip min/avg/max = 0.070/0.081/0.097 ms
+```
+
+Now let us connect container6 to the `local_alias` network with a different `network alias`.
````
can you put `container6` in back tics?

thaJeztah commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,129 @@ environment variables into a running container without significant effort and he
+those containers that are connected to that network can access the alias.
+
+In addition to the above features, multiple containers can share the same network-scoped
+alias within the same network. For example, lets launch `container7` in `isolated_nw` with
```
s/lets/let's/

thaJeztah commented:
@mavenugo can you add the new flag (`--net-alias`) and the changes in formats for `--link` to;

- https://github.com/docker/docker/blob/master/docs/reference/commandline/network_connect.md
- https://github.com/docker/docker/blob/master/docs/reference/run.md
- https://github.com/docker/docker/blob/master/man/docker-network-connect.1.md
- https://github.com/docker/docker/blob/master/man/docker-run.1.md

(I think this is only supported on `docker run`, or also on `docker create` and `docker start`?)

mavenugo commented:
thanks @thaJeztah. updated the docs. PTAL.

thaJeztah commented:
hmf, key servers are still down apparently :cry:

vdemeester commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,130 @@ environment variables into a running container without significant effort and he
+While `links` provide private name resolution that is localized within a container,
+the network-scoped alias provides a way for a container to be discovered by an
+alternate name by any other container within the scope of a particular network.
+Unlike the `link` alias, which is defined by the consumer of a service, the
```
Is *alias* needed here ?

```makrdown 
Unlike the `link`, which is defined by the consumer of a service, the
network-scoped alias is defined by the container that is offering the service
to the network.
```

mavenugo replied:
Yes. we are trying to differentiate the `alias` functionality provided by `link` and `--net-alias`. Infact we call it out in the docs as `alias` for link as well. So, it is better to be explicit here.

vdemeester replied:
Ah I see, between the local scope aliases (specified by `--link`) and thoses one. Maybe we could specify that ? :stuck_out_tongue_closed_eyes:

mavenugo replied:
Not really. `--link` always called it an alias : `--link=""  : Add link to another container (<name or id>:alias or <name or id>)` as in `https://github.com/docker/docker/blob/master/docs/reference/run.md` and many other places.

vdemeester replied:
Fair enough :+1:

vdemeester commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,130 @@ environment variables into a running container without significant effort and he
+alternate name by any other container within the scope of a particular network.
+Unlike the `link` alias, which is defined by the consumer of a service, the
+network-scoped alias is defined by the container that is offering the service
+to the network
```
Missed a period here `.`.

vdemeester commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,130 @@ environment variables into a running container without significant effort and he
+`container6` in this example now is aliased as `app` in network `isolated_nw` and
+as `scoped-app` in network `local_alias`.
+
+Let's try to reach these aliases from `container4` (which is connected to both these networks)
```
Missed a period here too `.`.

mavenugo replied:
added the missing `.` after the mention for `container5` below.

vdemeester commented on docs/userguide/networking/work-with-networks.md:
```diff
@@ -499,6 +499,130 @@ environment variables into a running container without significant effort and he
+When multiple containers share the same alias, name resolution to that alias will happen
+to one of the containers (typically the first container that is aliased). When the container
+that backs the alias goes down or disconnected from the network, the next container that
+backs the alias will be resolved.
```
It's not really related to the docs-review but the choice of one container over another is based on what ? the time it was created, the name, or something else ? :stuck_out_tongue_closed_eyes:

mavenugo replied:
In the current implementation, it is first-come first-serve. Which is mentioned in the above doc
`typically the first container that is aliased`.

But in a multi-host cluster environment, it is difficult to explain which is the first container (when multiple containers are started in parallel across the hosts).

vdemeester replied:
@mavenugo ok, thanks for explaining.

vdemeester commented:
@mavenugo you need to update `docs/reference/commandline/run.md` and `docs/reference/commandline/create.md` too ; and a few nits :wink:.

tiborvass commented:
Ping @vdemeester

vdemeester commented:
A tiny last comment but overall great `\o/`.

docs LGTM 🐍

mavenugo commented:
Windows failure is unrelated and am seeing the same failure in multiple PRs.

vdemeester commented:
@mavenugo yep :sob: 
ping @thaJeztah @jamtur01 @MHBauer @moxiegirl for one last look :wink:

thaJeztah commented:
ping @mavenugo looks like you missed the network-connect man page
https://github.com/docker/docker/blob/master/man/docker-network-connect.1.md

can you add that, then we should be ready to go!

mavenugo commented:
@thaJeztah I forgot to comment on that earlier. This particular file needs some updates for multiple options. (ip, ip6, link and alias). I thought of addressing it via the other issue opened for addressing the docs. Is that okay ?

thaJeztah commented:
Alright, let's do that :smile:

LGTM

calavera commented:
all :green_heart: merging! :tada:

mavenugo commented:
Thanks all. 
@aanand can breath now :)

vdemeester commented:
:tada:
